File: ui/src/api/i18n.js
import { I18n } from "i18n-js";
import en from "../locales/en.json";
import be from "../locales/be.json";
import ru from "../locales/ru.json";
import { getTelegramLanguage } from "./telegram";

const DEFAULT_LANGUAGE = "en";
const i18n = new I18n();

const loadLocales = () => {
    i18n.enableFallback = true;
    i18n.defaultLocale = DEFAULT_LANGUAGE;
    i18n.translations = {
        en: en,
        be: be,
        ru: ru
    };

    i18n.locale = getTelegramLanguage();
};

loadLocales();
export { i18n };

File: api/provider/translation/i18n.go
package translation

import (
	"encoding/json"

	"github.com/nicksnyder/go-i18n/v2/i18n"
	"golang.org/x/text/language"
)

type I18NTranslationProvider struct {
	bundle *i18n.Bundle
}

func (provider *I18NTranslationProvider) TranslateMessage(messageID string, locale string, templateData TemplateData) (string, error) {
	localizer := i18n.NewLocalizer(provider.bundle, locale)
	translation, tag, err := localizer.LocalizeWithTag(&i18n.LocalizeConfig{
		MessageID:    messageID,
		TemplateData: templateData,
	})

	if err != nil && tag == language.Und {
		return "", err
	}

	return translation, nil
}

func CreateI18NTranslationProvider() (TranslationProvider, error) {
	bundle := i18n.NewBundle(DEFAULT_LANGUAGE)
	bundle.RegisterUnmarshalFunc("json", json.Unmarshal)

	locales := []string{
		"locales/en.json",
		"locales/be.json",
		"locales/ru.json",
	}

	for _, locale := range locales {
		_, err := bundle.LoadMessageFile(locale)
		if err != nil {
			return nil, err
		}
	}

	provider := I18NTranslationProvider{
		bundle: bundle,
	}

	return &provider, nil
}
